core/runtime/execute_engine.cpp

```diff
@@ -168,27 +210,66 @@ std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intr
       auto dims = core::util::toVec(out_shape);
       auto type = util::TRTDataTypeToScalarType(compiled_engine->exec_ctx->getEngine().getTensorDataType(name.c_str()));
       outputs[pyt_idx] = std::move(at::empty(dims, {at::kCUDA}).to(type).contiguous());
+
+      // In cudagraphs mode, the allocated output buffers are stored for reuse
+      if (CUDAGRAPHS_MODE) {
+        compiled_engine->output_buffers[pyt_idx] = outputs[pyt_idx];
+      }
       compiled_engine->exec_ctx->setTensorAddress(name.c_str(), outputs[pyt_idx].data_ptr());
     }
   }
 
-  {
-    std::unique_ptr<torch::autograd::profiler::RecordProfile> enqueue_profiler_guard;
-    if (compiled_engine->profile_execution) {
-      enqueue_profiler_guard =
-          std::make_unique<torch::autograd::profiler::RecordProfile>(compiled_engine->enqueue_profile_path);
-    }
+  std::unique_ptr<torch::autograd::profiler::RecordProfile> enqueue_profiler_guard;
+
+  // nvinfer1::IExecutionContext::enqueue is not thread safe and we need a mutex for it.
+  std::unique_lock<std::mutex> lock(compiled_engine->mu);
+
+  if (!CUDAGRAPHS_MODE) {
+    // If not in cudagraphs mode, proceed with enqueueV3 as normal
     c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream(inputs[0].device().index());
```

Comment:
Can we take this oppurtunity to fix the default stream issue? I think we need to create a new stream for the enqueue operation

Reply:
Yes, I think in both the Python and C++ cases, we can use a non-default stream. I am not sure if we want to grab a new stream from the pool on every `forward` pass, or if we should store the stream after first inference call.

Reply:
`self.stream`
--> Synchronize at I/O boundaries
```python
stream = torch.cuda.Stream()
torch.cuda.set_stream(stream)
```

Reply:
```py
torch.cuda.synchronize()
with torch.cuda.stream(s1):
    C = torch.mm(A, A)
```